### PR TITLE
Lrng 2

### PR DESCRIPTION
**Pull request by dsman1823**

### CODE REVIEW

File: src/main/scala/com/example/App.scala [hunk 1]
```diff
@@ -1,40 +1,53 @@
...
-
-    reporters.foreach(_.report)
+    val questionsDs = Extractor.readCsv[Question]("Questions.csv")
+
+    val data = AnswerScoreModel.transformData(answersDs, questionsDs)
+    val Array(train, test) = data.randomSplit(Array(.8, .2))
+
+    val linearRegression = new LinearRegression()
+    val linearParams = new ParamGridBuilder().addGrid(linearRegression.elasticNetParam, Array(0.3, 0.6, 0.8)).build()
+    val linearModel = AnswerScoreModel.model(train, linearParams, linearRegression)
```
Comment (greglira) [inline src/main/scala/com/example/App.scala:43]: do you need to tune regularization param as well?

File: src/main/scala/com/example/App.scala [hunk 2]
```diff
@@ -1,40 +1,53 @@
...
+
+
+    val gbtRegressor = new GBTRegressor()
+    val gbtParams = new ParamGridBuilder().addGrid(gbtRegressor.featureSubsetStrategy, Array("auto", "all")).build()
```
Comment (greglira) [inline src/main/scala/com/example/App.scala:47]: very few params tuned here. Please see the attached learning notebook you can use as a reference.

[ML 08 - Tuning Bikeshare.scala.zip](https://github.com/temp-sprk-lrng/spark-lrng/files/3780706/ML.08.-.Tuning.Bikeshare.scala.zip)

File: src/main/scala/com/example/App.scala [hunk 3]
```diff
@@ -1,40 +1,53 @@
...
-      expertsReporter,
-      QuotedReporter(answersDs, 100),
-      SeasonalityReporter(questionsDs, answersDs)
-    )
-
-    reporters.foreach(_.report)
+    val questionsDs = Extractor.readCsv[Question]("Questions.csv")
+
+    val data = AnswerScoreModel.transformData(answersDs, questionsDs)
+    val Array(train, test) = data.randomSplit(Array(.8, .2))
```
Comment (greglira) [inline src/main/scala/com/example/App.scala]: No cross-validation, it is something worth trying

File: src/main/scala/com/example/analytics/AnswerScoreModel.scala [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+  def model(train: DataFrame,
+            params: Array[ParamMap],
+            regressor: PipelineStage) = {
+    val tokenized = new Tokenizer().setInputCol("body").setOutputCol("tokenized")
+
+    val stopWordFree = new StopWordsRemover()
+      .setStopWords(StopWordsRemover.loadDefaultStopWords("english"))
+      .setInputCol("tokenized")
+      .setOutputCol("free")
```
Comment (greglira) [inline src/main/scala/com/example/analytics/AnswerScoreModel.scala:42]: need better naming for columns here

File: src/main/scala/com/example/analytics/AnswerScoreModel.scala [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      .setPredictionCol("prediction")
+      .setMetricName("rmse")
+
+
+    val cv = new CrossValidator()
+      .setEstimator(pipeline)
+      .setEvaluator(rmseEvaluator)
+      .setEstimatorParamMaps(params)
+
+    cv.fit(train)
```
Comment (greglira) [inline src/main/scala/com/example/analytics/AnswerScoreModel.scala:64]: very poor feature engineering here. many more features (about user, text, context, even time) should be extracted and used in the model. please work MORE on that.

File: src/main/scala/com/example/analytics/AnswerScoreModel.scala [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    val stopWordFree = new StopWordsRemover()
+      .setStopWords(StopWordsRemover.loadDefaultStopWords("english"))
+      .setInputCol("tokenized")
+      .setOutputCol("free")
+
+    val word2vec = new Word2Vec().setInputCol("free").setOutputCol("word2vec").setVectorSize(10)
+
+    val assembler = new VectorAssembler()
+      .setInputCols(Array("contains_link", "code_len", "question_score", "word2vec"))
```
Comment (greglira) [inline src/main/scala/com/example/analytics/AnswerScoreModel.scala:47]: try scaling for the features

File: src/main/scala/com/example/App.scala [hunk 7]
```diff
@@ -1,40 +1,53 @@
...
+    val mae = new RegressionEvaluator().setMetricName("mae").evaluate(predictions)
+
+    log.info("**********************************")
+    log.info(s"rmse: $rmse")
+    log.info(s"mae: $mae")
+    log.info("**********************************")
+  }
+
   def main(args: Array[String]): Unit = {
     init(args)
```
Comment (greglira) [inline src/main/scala/com/example/App.scala:34]: I wouldn't overload main with all this logic, need better design for this

File: src/main/scala/com/example/App.scala [hunk 8]
```diff
@@ -1,40 +1,53 @@
...
+
+    val linearRegression = new LinearRegression()
+    val linearParams = new ParamGridBuilder().addGrid(linearRegression.elasticNetParam, Array(0.3, 0.6, 0.8)).build()
+    val linearModel = AnswerScoreModel.model(train, linearParams, linearRegression)
+
+
+    val gbtRegressor = new GBTRegressor()
+    val gbtParams = new ParamGridBuilder().addGrid(gbtRegressor.featureSubsetStrategy, Array("auto", "all")).build()
+    val gbtModel = AnswerScoreModel.model(train, gbtParams, gbtRegressor)
+
```
Comment (greglira) [inline src/main/scala/com/example/App.scala:49]: I would also try gradient boosting algos (xgboost, catboost, lightgbm).
Check this notebook as a reference:

**Review by greglira — CHANGES_REQUESTED** [review]
Need more work on this. Typical ML project (even for educational purposes) requires more work in terms of data exploration (visualization etc), feature engineering, using various algorithms, tuning.
I would highly recommend this course https://mlcourse.ai/ to complete or at least get yourself familiar with the ml basics. This will be important even for a data engineer, so he understands all the steps/phases in a typical ML project.
